lib/web/apiserver.go

```diff
@@ -2688,6 +2692,60 @@ func (h *Handler) WithClusterAuth(fn ClusterHandler) httprouter.Handle {
 	})
 }
 
+// ProvisionTokenHandler is a authenticated handler that is called for some existing Token
+type ProvisionTokenAuthedHandler func(w http.ResponseWriter, r *http.Request, p httprouter.Params, site reversetunnel.RemoteSite) (interface{}, error)
+
+// WithProvisionTokenAuth ensures that request is authenticated with a provision token.
+// Provision tokens, when used like this are invalidated as soon as used.
+// Doesn't matter if the underlying response was a success or an error.
+func (h *Handler) WithProvisionTokenAuth(fn ProvisionTokenAuthedHandler) httprouter.Handle {
+	return httplib.MakeHandler(func(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
+		ctx := r.Context()
+		logger := h.log.WithField("request", fmt.Sprintf("%v %v", r.Method, r.URL.Path))
+
+		creds, err := roundtrip.ParseAuthHeaders(r)
+		if err != nil {
+			logger.WithError(err).Warn("No auth headers.")
+			return nil, trace.AccessDenied("need auth")
+		}
+
+		if err := h.consumeTokenForAPICall(ctx, creds.Password); err != nil {
+			h.log.WithError(err).Warn("Failed to authenticate.")
+			return nil, trace.AccessDenied("need auth")
+		}
+
+		clusterName := p.ByName("site")
+		if clusterName == currentSiteShortcut {
+			res, err := h.GetProxyClient().GetClusterName()
+			if err != nil {
+				h.log.WithError(err).Warn("Failed to query cluster name.")
+				return nil, trace.Wrap(err)
+			}
+			clusterName = res.GetClusterName()
+		}
+
+		site, err := h.cfg.Proxy.GetSite(clusterName)
+		if err != nil {
+			h.log.WithError(err).WithField("cluster-name", clusterName).Warn("Failed to query site.")
+			return nil, trace.Wrap(err)
+		}
+
+		return fn(w, r, p, site)
+	})
+}
+
+func (h *Handler) consumeTokenForAPICall(ctx context.Context, tokenName string) error {
+	token, err := h.GetProxyClient().GetToken(ctx, tokenName)
```

Comment:
How do we differentiate between the token that was created for some other reason, e.g. a user token or a node join token?

Reply:
I'm not sure a new join method would fit "conceptually" tbh - join method is for agents joining the cluster while we just need to download the certs. Can we set some label or annotation or type on the provision token and require it here?

Reply:
And also, yeah, we should only query for specifically a provisioning token here.

Reply:
Agreed, it's stretching the "Join" concept

What do you think about adding an internal label where we say which endpoint the Token is available to call?
So, to call the `POST /webapi/sites/:site/sign` we would need a token with the following label
`teleport.internal/purpose: sign`
`teleport.internal/purpose: api/sign`
`teleport.internal/purpose: auth/sign`
?

Reply:
Something like this
https://github.com/gravitational/teleport/pull/14526/commits/225dcf475ee7765e3382a5868cd88c2aa5bdd799

Reply:
@marcoandredinis I wonder if we're overloading existing concepts trying to make them work for our use-case. What if we keep it simple and just require a label on the token e.g. `teleport.internal/purpose: database-mtls`? The UI would need to set it when creating the token and we would check for it here. WDYT?

Reply:
@marcoandredinis Actually, disregard my suggestion above. Thinking about it more, requiring database role in the token makes sense. You're joining a database agent which will require a token with "db" role in it, so requiring the same role in the token to generate database certificates makes sense also. Let's go with your suggestion.